File: Haskell/Ray.hs
-- File: Ray.hs

-- Author: Christopher Camargo
-- CMPS 112: Winter 2014

-- Ray module declaration.
module Ray 
( Ray (Ray)
, origin
, direction
, getIntersection
, getx
, gety
, getRay
) where

import Vec3

-- Ray has an origin, direction, and t parameter.
data Ray = Ray { origin :: Vec3, direction :: Vec3 } deriving (Show)

getIntersection :: Ray -> Float -> Vec3 -- Returns a vector representing an intersection point.
getIntersection ray t = add (origin ray) (mult (direction ray) t)

getx :: Int -> Float -- Get the x coordinate of a pixel. Using 50 degree fov.
getx x = (2.0 * (((fromIntegral x :: Float) + 0.5) * (1.0/512.0)) - 1.0) * (tan (pi * 0.5 * 50.0 / 180.0))

gety :: Int -> Float -- Get the y coordinate of a pixel. Using 50 degree fov.
gety y = (1.0 - 2.0 * (((fromIntegral y :: Float) + 0.5) * (1.0/512.0))) * (tan (pi * 0.5 * 50.0 / 180.0))

getRay :: Int -> Int -> Ray -- Get a ray through the x and y pixel.
getRay x y = Ray (Vec3 0.0 0.0 0.0) ( normalize (Vec3 (getx x) (gety y) (-1.0)))

File: Java/Ray.java
// File: Ray.java

// Author: Christopher Camargo
// CMPS 112: Winter 2014

public class Ray {
	private Vec3 origin; // Ray point of origin.
	private Vec3 direction; // Direction of Ray.
	private float t; // Parameter for this Ray.

	// Main constructor for a Ray.
	public Ray(Vec3 o, Vec3 d) {
		origin = o; 
		direction = d;
	}

	// Returns a vector representing an intersection point with a shape.
	public Vec3 getIntersection() {
		return origin.add(direction.mult(t));
	}

	// Sets the parameter t of a Ray.
	public void setT(float t) {
		this.t = t;
	}

	// Returns the origin of a Ray.
	public Vec3 origin() {
		return origin;
	}

	// Returns the direction of a Ray.
	public Vec3 direction() {
		return direction;
	}

	// Returns the parameter t of a Ray.
	public float t() {
		return t;
	}
}